Format code with black and isort

pquentin opened the pull request:
This is a proposal to switch to black (and isort, as black does not sort imports yet.)

@danielmitterdorfer said this in #1276:

> black is on our radar but I fear that we won't have much capacity at the moment to work on it.

I'm not sure whether it was the actual pull request that was the issue, or the work needed to adapt (eg. everyone will have to configure their editors). My experience is that black allows to focus on the interesting parts of the code, which is highly valuable. Here's what Mike Bayer, creator and maintainer of SQLAlchemy [says about black](https://twitter.com/zzzeek/status/1398662789078650884):

> I would say that I can't think of any single tool in my entire programming career that has given me a bigger productivity increase by its introduction.   commits like https://github.com/sqlalchemy/sqlalchemy/commit/ebbbac0a76b3327a829864afb26ee1b7ff1dc780, Black did 90% of the work for me that previously we'd spend days formatting by hand

This time, lint passes (I was confused because lint fails on master with Python 3.9). The actual black run is in its own commit, so this pull request is easy to rebase. We can also tell git to ignore the black commit in `git blame`.

pquentin commented on benchmarks/driver/parsing_test.py:
```diff
@@ -25,196 +25,140 @@
-    group="parse",
-    warmup="on",
-    warmup_iterations=10000,
-    disable_gc=True
```
black collapes this into a single line, but we can prevent it if needed by adding a trailing comma here

danielmitterdorfer replied:
I'd personally prefer that we keep the current formatting here.

pquentin replied:
done

danielmitterdorfer commented:
@elasticmachine test this please

danielmitterdorfer commented on benchmarks/__init__.py:
```diff
@@ -6,7 +6,7 @@
 # You may obtain a copy of the License at
 #
-#	http://www.apache.org/licenses/LICENSE-2.0
+# 	http://www.apache.org/licenses/LICENSE-2.0
```
Yes, that's ok.

danielmitterdorfer commented on Makefile:
```diff
@@ -87,7 +87,13 @@ tox-env-clean:
 lint: check-venv
-	@find esrally benchmarks scripts tests it -name "*.py" -exec $(VEPYLINT) -j0 -rn --load-plugins pylint_quotes --rcfile=$(CURDIR)/.pylintrc \{\} +
+	@find esrally benchmarks scripts tests it -name "*.py" -exec $(VEPYLINT) -j0 -rn --rcfile=$(CURDIR)/.pylintrc \{\} +
+	@black --check esrally benchmarks scripts tests it
```
```suggestion
	@. $(VENV_ACTIVATE_FILE); black --check esrally benchmarks scripts tests it
```

pquentin replied:
Should I use VEBLACK for consistency with VEPYLINT above?

pquentin replied:
Or use `VENV_ACTIVATE_FILE` for pylint too?

danielmitterdorfer replied:
We only use `VEPYLINT` above because of `find` so let's stick to `VENV_ACTIVATE_FILE` here.

danielmitterdorfer commented on Makefile:
```diff
@@ -87,7 +87,13 @@ tox-env-clean:
-	@find esrally benchmarks scripts tests it -name "*.py" -exec $(VEPYLINT) -j0 -rn --load-plugins pylint_quotes --rcfile=$(CURDIR)/.pylintrc \{\} +
+	@find esrally benchmarks scripts tests it -name "*.py" -exec $(VEPYLINT) -j0 -rn --rcfile=$(CURDIR)/.pylintrc \{\} +
+	@black --check esrally benchmarks scripts tests it
+	@isort --check esrally benchmarks scripts tests it
```
```suggestion
	@. $(VENV_ACTIVATE_FILE); isort --check esrally benchmarks scripts tests it
```

danielmitterdorfer commented on Makefile:
```diff
@@ -87,7 +87,13 @@ tox-env-clean:
+	@isort --check esrally benchmarks scripts tests it
+
+format: check-venv
+	@black esrally benchmarks scripts tests it
```
```suggestion
	@. $(VENV_ACTIVATE_FILE); black esrally benchmarks scripts tests it
```

danielmitterdorfer commented on Makefile:
```diff
@@ -87,7 +87,13 @@ tox-env-clean:
+
+format: check-venv
+	@black esrally benchmarks scripts tests it
+	@isort esrally benchmarks scripts tests it
```
```suggestion
	@. $(VENV_ACTIVATE_FILE); isort esrally benchmarks scripts tests it
```

danielmitterdorfer reviewed (commented):
I've triggered a CI build but it failed with:

```
Makefile:90: recipe for target 'lint' failed
/bin/bash: black: command not found
make: *** [lint] Error 127
```

The reason is that `black` and `isort` are only available inside the virtualenv but our build scripts always activate the virtualenv lazily (i.e. only when needed). I left suggestions in `Makefile` that address this.

danielmitterdorfer commented:
> I'm not sure whether it was the actual pull request that was the issue, or the work needed to adapt (eg. everyone will have to configure their editors). 

It's actually a combination of both. :) Thanks for raising the PR to move this forward.

pquentin commented:
I fixed the `pytest.mark.benchmark` formatting, applied your suggestions to the Makefile, explained how to configure black and isort, and fixed the conflict.

Please take another look

danielmitterdorfer commented:
@elasticmachine test this please

danielmitterdorfer commented on esrally/driver/__init__.py:
```diff
@@ -16,4 +16,11 @@
 
 # expose only the minimum API
-from .driver import DriverActor, PrepareBenchmark, PreparationComplete, StartBenchmark, BenchmarkComplete, TaskFinished
+from .driver import (
```
It seems that it is configurable how multi-line imports are rendered. Can we please use `11 - Backslash Grid` for such cases (see their [docs](https://pycqa.github.io/isort/docs/configuration/multi_line_output_modes/))?

pquentin replied:
While black does not *sort* imports, it does reformat them: we can only use the black mode in isort, not the other ones. black does this to minimize diffs.

danielmitterdorfer commented on pyproject.toml:
```diff
@@ -0,0 +1,6 @@
+[tool.black]
+line-length = 140
```
Let's keep the 140 characters for now.

danielmitterdorfer reviewed (commented):
Thanks for iterating! I had a closer look now and left some detailed comments. Some high-level observations:

* It seems that `black` has formatted the code with a different line-length limit than what's configured at the moment. For example, I manually reformatted the runner registration code in `runner.py` to match the prior format and ran `make format` again but this time it did not change the code. Can you please reformat the code so it respects the 140 character limit?
* The rendering of multi-line imports should be configured as backslash grid.
* We should add a note in our development docs about IDE setup.

Apart from that the changes look already pretty good!

pquentin commented on esrally/driver/runner.py:
```diff
@@ -52,7 +52,11 @@ def register_default_runners():
     register_runner(track.OperationType.Composite, Composite(), async_runner=True)
     register_runner(track.OperationType.SubmitAsyncSearch, SubmitAsyncSearch(), async_runner=True)
-    register_runner(track.OperationType.GetAsyncSearch, Retry(GetAsyncSearch(), retry_until_success=True), async_runner=True)
+    register_runner(
```
Thank you for the careful review! I tried a few line length settings, but didn't realize they were "sticky". I fixed this by rebasing and applied `make format` on the original code, which fixed it, as you mentioned.

(It took me some time because I had to work around https://github.com/psf/black/issues/2321.)

pquentin commented on CONTRIBUTING.md:
```diff
@@ -54,16 +54,20 @@ Once your changes and tests are ready to submit for review:
-2. Sign the Contributor License Agreement
+2. Format your changes
+
+    Also make sure that lint is passing by running `make lint`. To format your code, run `make format`. Consider using editor integrations to do it automatically: you'll need to configure [black](https://black.readthedocs.io/en/stable/integrations/editors.html) and [isort](https://github.com/PyCQA/isort/wiki/isort-Plugins).
```
Thanks, done

pquentin commented:
We're down from 17k new lines to 11k new lines thanks to the line length fix. 🎉 This is ready for another round of review!

pquentin commented on tests/tracker/index_test.py:
```diff
@@ -82,83 +67,36 @@ def test_extract_index_create(client):
-                }
-            }
-        }
+        "geodata": {"mappings": {}, "settings": {"index": {"number_of_shards": "{{number_of_shards | default(1)}}"}}},
```
With the large line length, we have many examples like this across the codebase that are arguably more difficult to read, I can again try to track them down and add trailing commas. This can probably be done semi-automatically by looking at the git diff for "multiple lines that got collapsed to a single one".

danielmitterdorfer commented on tests/metrics_test.py:
```diff
@@ -1846,381 +1692,198 @@ def test_as_flat_list(self):
-            "name": "throughput",
-            "task": "index #1",
-            "operation": "index",
-            "value": {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/metrics_test.py:
```diff
@@ -1846,381 +1692,198 @@ def test_as_flat_list(self):
-            "name": "service_time",
-            "task": "index #1",
-            "operation": "index",
-            "value": {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/metrics_test.py:
```diff
@@ -1846,381 +1692,198 @@ def test_as_flat_list(self):
-            "name": "throughput",
-            "task": "search #2",
-            "operation": "search",
-            "value": {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/metrics_test.py:
```diff
@@ -1846,381 +1692,198 @@ def test_as_flat_list(self):
-            }
-        }, select(metric_list, "throughput", operation="search"))
-
-        self.assertEqual({
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/metrics_test.py:
```diff
@@ -1846,381 +1692,198 @@ def test_as_flat_list(self):
-            }
-        }, select(metric_list, "service_time", operation="search"))
-
-        self.assertEqual({
```
Can we please keep the current formatting here? I'll stop my individual comments at this point but this applies similarly to all nested structures below.

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -274,55 +323,49 @@ def test_generates_log_file(self, file_mock, es):
         segment_stats = telemetry.SegmentStats("/var/log", es)
         segment_stats.on_benchmark_stop()
         es.cat.segments.assert_called_with(index="_all", v=True)
-        file_mock.assert_has_calls([
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -274,55 +323,49 @@ def test_generates_log_file(self, file_mock, es):
         clients = {"default": Client(), "cluster_b": Client()}
         cfg = create_config()
         metrics_store = metrics.EsMetricsStore(cfg)
-        telemetry_params = {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -408,18 +451,12 @@ def test_stores_default_ccr_stats(self, metrics_store_put_doc):
+        shard_metadata = {"cluster": "remote", "index": follower_index}
 
         metrics_store_put_doc.assert_called_with(
-            {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -505,34 +543,22 @@ def test_stores_default_ccr_stats_many_shards(self, metrics_store_put_doc):
         recorder = telemetry.CcrStatsRecorder("remote", client, metrics_store, 1)
         recorder.record()
 
-        shard_metadata = [
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -651,21 +676,17 @@ def test_stores_filtered_ccr_stats(self, metrics_store_put_doc):
 
-        metrics_store_put_doc.assert_has_calls([
-            mock.call(
-                {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -790,100 +803,61 @@ def test_stores_multi_index_multi_shard_stats(self, metrics_store_put_doc):
-                        "total_time_in_millis": 100
+                        "total_time_in_millis": 100,
                     },
-                    {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -790,100 +803,61 @@ def test_stores_multi_index_multi_shard_stats(self, metrics_store_put_doc):
             },
             "index2": {
                 "shards": [
-                    {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -790,100 +803,61 @@ def test_stores_multi_index_multi_shard_stats(self, metrics_store_put_doc):
+        )
         recorder.record()
 
-        metrics_store_put_doc.assert_has_calls([
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -987,437 +948,179 @@ def test_missing_properties_shard_stats(self, metrics_store_put_doc):
-                "total_size": 279900,
-                "open_count": 274,
-                "close_count": 274,
-                "contiguous_bytes_read": {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -987,437 +948,179 @@ def test_missing_properties_shard_stats(self, metrics_store_put_doc):
-                "total_size": 356841728759,
-                "open_count": 174,
-                "close_count": 174,
-                "contiguous_bytes_read": {
```
Can we please keep the current formatting here?

danielmitterdorfer commented on tests/telemetry_test.py:
```diff
@@ -987,437 +948,179 @@ def test_missing_properties_shard_stats(self, metrics_store_put_doc):
-                    "time_in_nanos": 0
-                },
-                "forward_seeks": {
-                    "small": {
```
Can we please keep the current formatting here? Similarly for all nested structures below.

danielmitterdorfer reviewed (commented):
Thanks for iterating. While I left a lot of comments it boils down to the following issues:

* Elasticsearch API requests and responses in tests should not be reformatted. Especially when these structures are nested they're hard to grasp visually or to modify. This affects especially `runner_test.py`, `metrics_test.py` and `telemetry_test.py`. If it's possible with black I'd probably even go as far as to exclude these files from formatting entirely (although there are places where black has improved formatting IMHO).
* In other tests we intentionally format the code in a way that the visual structure matches the logical structure so it's easier to reason. I'm specifically thinking of `AllocatorTests` here. Also, here I'd opt that we keep the current formatting.
* IMHO dict literals containing more than a few items (~ 2-3) are easier to read when each item is on its own line.

danielmitterdorfer commented:
@elasticmachine test this please

pquentin commented:
Wow, you really left 200+ comments! I would have been happy with only your main review. I'll fix everything, but not sure when it will be done, maybe next week.

> If it's possible with black I'd probably even go as far as to exclude these files from formatting entirely (although there are places where black has improved formatting IMHO).

It's certainly possible to give to the `black` tool a specific list of files in `make lint` and `make format`, but then everyone would have to try to configure their editor to do that, so I'd rather avoid that and just put the hard work now. (Even if I have to admit this is more involved than I would have thought.)

danielmitterdorfer commented:
> Wow, you really left 200+ comments! I would have been happy with only your main review. I'll fix everything, but not sure when it will be done, maybe next week.

Honestly, I was also surprised that it added up to that many comments but I wanted not only mention the high-level feedback but also provide specific guidance around what formatting I think needs to be kept. Nevertheless I suggest you focus on my high-level comments; no need to address each and every one of them.

>>    If it's possible with black I'd probably even go as far as to exclude these files from formatting entirely (although there are places where black has improved formatting IMHO).

> It's certainly possible to give to the black tool a specific list of files in make lint and make format, but then everyone would have to try to configure their editor to do that, so I'd rather avoid that and just put the hard work now. (Even if I have to admit this is more involved than I would have thought.)

Good point, let's not exclude certain files then. The intention behind my proposal was to avoid that you need to manually take care of formatting of the many dicts that we use in the tests in these files.

pquentin commented:
I addressed comments in all files with less than 1k lines, will do the rest tomorrow.

danielmitterdorfer reviewed (commented):
Thanks for the changes so far. I did a quick check; looks much better!

danielmitterdorfer commented:
@elasticmachine test this please

pquentin commented on tests/driver/runner_test.py:
```diff
@@ -4824,54 +3810,33 @@ async def test_search_after_with_pit(self, es):
             "results-per-page": 2,
-            "body": {
-                "sort": [{"timestamp": "asc", "tie_breaker_id": "asc"}],
-                "query": {"match-all": {}}
```
Should this be `match_all`?

danielmitterdorfer replied:
Yes, it should be.

pquentin replied:
Thanks, done in https://github.com/elastic/rally/pull/1281/commits/e57e08dcc9ec8aeade701e33cd13c0cc8594889e

pquentin commented on tests/driver/runner_test.py:
```diff
@@ -5345,23 +4209,27 @@ async def test_executes_tasks_in_specified_order(self, es):
+        self.assertEqual(
+            [
+                "initial-call",
+                # concurrent
```
This isn't possible. What about a `stream-a and stream-b are concurrent` comment?

danielmitterdorfer replied:
Ok, that works for me.

pquentin commented on tests/driver/runner_test.py:
```diff
@@ -344,50 +300,36 @@ class BulkIndexRunnerTests(TestCase):
         bulk = runner.BulkIndex()
 
         bulk_params = {
-            "body": "action_meta_data\n" +
```
Not easily: those are +, not commas.

I can use a `build_bulk_params` helper:

```
        bulk_params = {
            "body": _build_bulk_body(
                "action_meta_data",
                "index_line",
                "action_meta_data",
                "index_line",
                "action_meta_data",
                "index_line",
            )
        }
```

Would that work for you?

danielmitterdorfer replied:
Yes, that'd be fine.

danielmitterdorfer commented:
@elasticmachine ok to test

pquentin commented:
Thanks! I'm finally ready for another iteration. I marked all comments are resolved as it was the only to keep track of them, but I can unresolve them if you want me to.

When the pull request looks good to you, I'd like to rebase it on top of master to avoid messing up too much with the history.

danielmitterdorfer reviewed (approved):
Thanks for iterating and all the work you've put into this! The changes look fine to me now (I left one very minor nit). As this PR has a rather broad impact, I've asked @dliappis to have a look as well but I think we're close to merging this.

> When the pull request looks good to you, I'd like to rebase it on top of master to avoid messing up too much with the history.

That would be great. Ideally we'd have two commits: One that adds the configuration and one for the reformatting. Then we can use an [`ignoreRevsFile`](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html#avoiding-ruining-git-blame) to only ignore the latter commit. I propose that you wait with this step until @dliappis had a chance to review it (no worries, that should be quicker than my review ;)).

pquentin commented:
(The tests ran fine but GitHub does not know it.)

dliappis reviewed (approved):
Thank you for this massive work!

This worked fine for me and is really useful. I tested it both via the IDE and `make format`.

A few comments:

1. Black seems to use some unicode code points (like https://codepoints.net/U+1F4A5) that are not necessarily part of all fonts. For example this is how it shows up on my terminal using the monospace font:

    ![image](https://user-images.githubusercontent.com/1754575/123936196-9fdbb100-d99d-11eb-927e-a241d5e597ba.png)

    Whereas it should be: `Oh no! 💥 💔 💥`
    After a quick glance of the config settings, I didn't find a way to stop it from doing that (or advice on what font to use for that matter). It's not a big problem, we are mostly interested in how it looks from the CI/Jenkins console, so if it shows up ok there, I am good.

2. For users who forget to run `make lint` before pushing, I think in a follow-up PR we could add a new `check-all` Makefile target that is easy to remember and run locally before pushing. We could include this in the contributing guide. This will uncover linting or (if lint passes) other issues with unit tests and IT tests. Right now we individually have to run `make lint`, `make test` and `make it`.

danielmitterdorfer commented:
Thank you for your review @dliappis!

@pquentin with two approvals on this PR we should be good to move forward. Can you please, as a last step before we merge this, push this as two commits as I've described above? As soon as this is done and the CI checks ran again, I'll merge this. Once more, thank you very much for all the effort that you put into it!

pquentin commented:
@dliappis Thanks for the review!

1. I currently use Monaco on macOS, and am not familiar with emoji handling on Linux, but I think you need a fallback emoji font such as noto-fonts-emoji. I have ordered a Linux laptop, so I might face the same issue at some point! This does display correctly in the CI, though:

   ![Screenshot 2021-06-30 at 14 21 02](https://user-images.githubusercontent.com/42327/123944692-6eb7ac80-d9ae-11eb-9dfd-a9a71ce66953.png)

   Another option is to use `--quiet` but I think the emojis are a nice visual way to see success (when they work!).
       
2. Sure, I'll add a `check-all` command. But I really recommand to configure black and isort in your editor (even though black is slow on the largest files in the repo).

@danielmitterdorfer Sure, I'll do that. I'll also be happy to help with rebasing any outstanding pull requests that will be affected by this.

pquentin commented:
@danielmitterdorfer Done!

danielmitterdorfer commented:
And now that CI has passed, it's merged. Thanks for your massive effort and persistence @pquentin!

pquentin commented:
🎉

Thank _you_! I was happy with my initial commits, and had no idea of the value of going through every single change like you did. You estimated the effort to do this much more accurately than me, even though I did that multiple times in the past. I'm really impressed.
